desc/magnetic_fields.py

```diff
@@ -1094,3 +1129,614 @@ def odefun(rpz, s):
+    modes_R : array-like, shape(k,2)
+        poloidal and toroidal mode numbers [m,n] for R_lmn.
+    modes_Z : array-like, shape(k,2)
+        mode numbers associated with Z_lmn, defaults to modes_R
+    NFP : int
+        number of field periods
+    sym : bool
+        whether to enforce stellarator symmetry for the toroidal surface.
+        Default is "auto" which enforces if modes are symmetric.
+        If True, non-symmetric modes will be truncated.
+    rho : float [0,1]
+        flux surface label for the toroidal surface
+    name : str
+        name for this field
+    check_orientation : bool
+        ensure that this surface has a right handed orientation. Do not set to False
+        unless you are sure the parameterization you have given is right handed
+        (ie, e_theta x e_zeta points outward from the surface).
+
+    """
+
+    _io_attrs_ = (
+        _MagneticField._io_attrs_
+        + FourierRZToroidalSurface._io_attrs_
+        + ["_surface_grid", "_Phi_mn", "_I", "_G"]
+    )
+
+    def __init__(
+        self,
+        Phi_mn=np.array([0.0]),
+        modes_Phi=np.array([[0, 0]]),
+        I=0,
+        G=0,
+        sym_Phi="auto",
+        R_lmn=None,
+        Z_lmn=None,
+        modes_R=None,
+        modes_Z=None,
+        NFP=1,
+        sym="auto",
+        rho=1,
+        name="",
+        check_orientation=True,
+    ):
+        self._Phi_mn = Phi_mn
+
+        Phi_mn, modes_Phi = map(np.asarray, (Phi_mn, modes_Phi))
+
+        assert np.issubdtype(modes_Phi.dtype, np.integer)
+
+        M_Phi = np.max(abs(modes_Phi[:, 0]))
+        N_Phi = np.max(abs(modes_Phi[:, 1]))
+
+        self._M_Phi = M_Phi
+        self._N_Phi = N_Phi
+
+        if sym_Phi == "auto":
+            if np.all(
+                Phi_mn[np.where(sign(modes_Phi[:, 0]) == sign(modes_Phi[:, 1]))] == 0
+            ):
+                sym_Phi = "sin"
+            elif np.all(
+                Phi_mn[np.where(sign(modes_Phi[:, 0]) != sign(modes_Phi[:, 1]))] == 0
+            ):
+                sym_Phi = "cos"
+            else:
+                sym_Phi = False
+            # catch case where only (0,0) mode is given as 0
+            if np.all(Phi_mn == 0.0) and np.all(modes_Phi == 0):
+                sym_Phi = "cos"
+        self._sym_Phi = sym_Phi
+        self._Phi_basis = DoubleFourierSeries(M=M_Phi, N=N_Phi, NFP=NFP, sym=sym_Phi)
+
+        assert np.isscalar(I), "I must be a scalar"
+        assert np.isscalar(G), "G must be a scalar"
+        self._I = I
+        self._G = G
+
+        super().__init__(
+            R_lmn, Z_lmn, modes_R, modes_Z, NFP, sym, rho, name, check_orientation
+        )
+
+    @property
+    def I(self):  # noqa: E743
+        """Net current linking the plasma and the surface toroidally."""
+        return self._I
+
+    @I.setter
+    def I(self, new):  # noqa: E743
+        assert np.isscalar(new), "I must be a scalar"
+        self._I = new
```

Comment:
we may also want to cast these to floats, doing AD with integers can give unexpected results